### DISCUSSION TITLE
Problem fetching and pulling data from GitHub

### DISCUSSION THREAD

**Discussion by Chaxiraxi** [atom/atom#22668] [Q&A]
# Problem fetching and pulling data from GitHub
Hi !

I have a problem when I want to pull and fetch my remote repo to my local repo.
It works fine in cli, and correctly fetch and pull my repo, but when I do it in Atom, it simply doesn't work and give me this error:
`fatal: couldn't find remote ref refs/heads/main`

Someone can tell me why it goes wrong ?
PS: It doesn't work with any repo, and I can commit and push in Atom without problems btw

**Comment by fleexerdesign** [comment]
I am having the same problem. But getting this message "fatal: protocol error: bad line length 2"

**Reply by andrerossidc** [reply]
I have exactly the same problem @fleexerdesign .

**Reply by kazhashimoto** [reply]
I am having the same problem on Atom 1.58.0 macOS Big Sur 11.5.
Git trace log says: Atom built-in git uses v2.26.2.
$ git config --global trace2.normalTarget ~/log.normal

On the other hand /usr/bin/git (pre-installed app in macOS) is v2.30.1 (Apple Git-130). This CLI git works fine.
Here is my workaround:

I replaced git executable files below with symbolic links to /usr/bin/git (v2.30.1)

/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git
/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git


![git-trace_edit](https://user-images.githubusercontent.com/50348276/129220145-0e418a3c-8711-4750-8b24-bf7573b58ecd.png)
![git-error-fix_edit](https://user-images.githubusercontent.com/50348276/129220059-89aff469-ecec-4567-91cc-bb3c3e0ee0dc.png)

**Reply by qubist** [reply]
This didn't work for me—it only disabled the git and GitHub panes in Atom and caused an error when trying to disable and reenable the `github` plugin

**Reply by ragumix** [reply]
@kazhashimoto it helped me! Thanks a lot!

**Reply by Simple-ID** [reply]
helped me too tx

**Reply by drancope** [reply]
Great, helped a lot

**Reply by qubist** [reply]
I've made an issue about this: https://github.com/atom/github/issues/2720#issuecomment-930277356

**Reply by isidoramorel** [reply]
Also worked for me, thanks!

**Reply by ph11be** [reply]
seems to have fixed it for me too....thank you!!

**Reply by DesertRider** [reply]
Worked for me too!

**Reply by lind1022** [reply]
I'm wondering how you managed to create symlink to "/usr/bin" which is protected by Mac's SIP?
I did the same but got error

`ln: /usr/bin/git: Operation not permitted`

**Reply by lind1022** [reply]
Hi @kazhashimoto, are you able to share how you created symlink to "/usr/bin" which is protected by SIP?

**Reply by kazhashimoto** [reply]
Hi @lind1022,
Here is my terminal session:
```
$ pwd
/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin
$ 
$ ln -s /usr/bin/git
$ ls -ln git
lrwxr-xr-x  1 501  20  12 11 18 11:55 git -> /usr/bin/git
$ 
```
I have confirmed that my Mac's SIP status is enabled and node_modules/dugite/git/bin directory is writable by my user id (501).
```
$ csrutil status
System Integrity Protection status: enabled.
$ ls -ldn $(pwd)
drwxr-xr-x@ 4 501  20  128 11 18 11:55 /Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin
```

**Reply by lind1022** [reply]
> Hi @lind1022, Here is my terminal session:
> 
> ```
> $ pwd
> /Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin
> $ 
> $ ln -s /usr/bin/git
> $ ls -ln git
> lrwxr-xr-x  1 501  20  12 11 18 11:55 git -> /usr/bin/git
> $ 
> ```
> 
> I have confirmed that my Mac's SIP status is enabled and node_modules/dugite/git/bin directory is writable by my user id (501).
> 
> ```
> $ csrutil status
> System Integrity Protection status: enabled.
> $ ls -ldn $(pwd)
> drwxr-xr-x@ 4 501  20  128 11 18 11:55 /Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin
> ```

That resolved my issue, thank you so much!!!

**Comment by tarkh** [comment]
I'm having similar problem.
Arch Linux, when trying to fetch repo updates, getting: "fatal: protocol error: bad line length 2"
While doing `git pull` from CLI works fine.
Atom 1.57.0-1

At the moment of error, Atom logs says this:
```
<embedded>:14 Uncaught (in promise) Error
    at new GitError (/usr/share/atom/resources/app/static/<embedded>:14)
    at /usr/share/atom/resources/app/static/<embedded>:14
GitError @ <embedded>:14
(anonymous) @ <embedded>:14
```

**Comment by realizelol** [comment]
Installing homebrew git and using this script helped me. TY @kazhashimoto

```
#!/bin/bash

if [[ ! -L "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git" ]]; then
  rm -f "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git"
  ln -s /usr/local/bin/git "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git"
fi

if [[ ! -L "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git" ]]; then
  rm -f "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git"
  ln -s /usr/local/bin/git "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git"
fi
```

**Reply by tmchow** [reply]
> Installing homebrew git and using this script helped me. TY @kazhashimoto
> 
> ```
> #!/bin/bash
> 
> if [[ ! -L "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git" ]]; then
>   rm -f "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git"
>   ln -s /usr/local/bin/git "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git"
> fi
> 
> if [[ ! -L "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git" ]]; then
>   rm -f "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git"
>   ln -s /usr/local/bin/git "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git"
> fi
> ```

Unfortunately this didn't work for me.  After installing git via homebrew and running that script, I'm getting this different error when trying to clone:

![image](https://user-images.githubusercontent.com/517103/130902328-62b58eb0-3d03-40af-9c16-f90be0c75f16.png)

Then I realized I don't have git binaries at `/usr/local/bin/git`.  Homebrew installs it to `/opt/homebrew/bin/git`.

So I then modified the script to be:

```
#!/bin/bash

if [[ ! -L "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git" ]]; then
  mv "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git" "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git.bak" 
  ln -s /opt/homebrew/bin/git "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/bin/git"
fi

if [[ ! -L "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git" ]]; then
  mv "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git" "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git.bak"
  ln -s /opt/homebrew/bin/git "/Applications/Atom.app/Contents/Resources/app.asar.unpacked/node_modules/dugite/git/libexec/git-core/git"
fi
```

and it now works

**Reply by NinjaCheetah** [reply]
Unfortunately neither of these scripts helped, in fact they completely killed the GitHub panel and made the Git panel no longer recognize my folders as repositories. Had to reinstall Atom to get it working again. (I verified that I do have `git` at `/usr/local/bin/git`) Strangely no error messages.

[a comment by an automated account is not shown]

**Comment by Tanapl02** [comment]
Thanks.
